pwndbg/dbg/lldb.py

```diff
@@ -9,22 +13,363 @@
+    return value.error.description and "optimized out" in value.error.description
+
+
+class LLDBType(pwndbg.dbg_mod.Type):
+    def __init__(self, inner: lldb.SBType):
+        self.inner = inner
+
+    @property
+    @override
+    def alignof(self) -> int:
+        return self.inner.GetByteAlign()
+
+    @property
+    @override
+    def code(self) -> pwndbg.dbg_mod.TypeCode:
+        return map_type_code(self.inner)
+
+    @override
+    def fields(self) -> List[pwndbg.dbg_mod.TypeField] | None:
+        fields = self.inner.get_fields_array()
+        return (
+            [
+                pwndbg.dbg_mod.TypeField(
+                    field.bit_offset,
+                    field.name,
+                    LLDBType(field.type),
+                    self,
+                    0,  # TODO: Handle fields for enum types differently.
+                    False,
+                    False,  # TODO: Handle base class members differently.
+                    field.bitfield_bit_size if field.is_bitfield else field.type.GetByteSize(),
+                )
+                for field in fields
+            ]
+            if len(fields) > 0
+            else None
+        )
+
+    @override
+    def array(self, count: int) -> pwndbg.dbg_mod.Type:
+        return LLDBType(self.inner.GetArrayType(count))
+
+    @override
+    def pointer(self) -> pwndbg.dbg_mod.Type:
+        return LLDBType(self.inner.GetPointerType())
+
+    @override
+    def strip_typedefs(self) -> pwndbg.dbg_mod.Type:
+        t = self.inner
+        while t.IsTypedefType():
+            t = t.GetTypedefedType
+
+        return LLDBType(t)
+
+    @override
+    def target(self) -> pwndbg.dbg_mod.Type:
+        t = self.inner.GetPointeeType()
+        if not t.IsValid():
+            raise pwndbg.dbg_mod.Error("tried to get target type of non-pointer type")
+
+        return LLDBType(t)
+
+
+class LLDBValue(pwndbg.dbg_mod.Value):
+    def __init__(self, inner: lldb.SBValue):
+        self.inner = inner
+
+    @property
+    @override
+    def address(self) -> pwndbg.dbg_mod.Value | None:
+        addr = self.inner.AddressOf()
+        return LLDBValue(addr) if addr.IsValid() else None
+
+    @property
+    @override
+    def is_optimized_out(self) -> bool:
+        return _is_optimized_out(self.inner)
+
+    @property
+    @override
+    def type(self) -> pwndbg.dbg_mod.Type:
+        assert not self.is_optimized_out, "tried to get type of optimized-out value"
+
+        return LLDBType(self.type)
+
+    @override
+    def dereference(self) -> pwndbg.dbg_mod.Value:
+        deref = self.inner.Dereference()
+
+        if not deref.IsValid():
+            raise pwndbg.dbg_mod.Error("could not dereference value")
+
+        return LLDBValue(deref)
+
+    @override
+    def string(self) -> str:
+        addr = self.inner.unsigned
+        error = lldb.SBError()
+
+        # Read strings up to 4GB.
+        last_str = None
+        buf = 256
+        for i in range(8, 33):
```

Comment:
Kinda unknown why `8, 33` - probably to account for reading exactly 4GB or sth? idk, but needs either a comment, or just a better range, or idk

Reply:
It covers the range 2^8 (the `256` in `buf`) to 2^32 (the 4GB in the comment), inclusive, since we double `buf` after every iteration to check for larger strings. It doesn't have to be `range(8, 33)`, but I think that's a little clearer in intent than `range(25)`. Maybe a `# log2(256) = 8, log2(4GB) = 32` or `# range(log2(256), log2(4GB) + 1)` comment could clear that up?